source/common/common/utility.cc

```diff
@@ -20,8 +21,116 @@
 #include "spdlog/spdlog.h"
 
 namespace Envoy {
+
+namespace {
+
+class SubsecondConstantValues {
+public:
+  const std::string PLACEHOLDER{"?????????"};
+  const std::regex PATTERN{"%([1-9])?f", std::regex::optimize};
+};
+
+typedef ConstSingleton<SubsecondConstantValues> SubsecondConstants;
+
+} // namespace
+
 std::string DateFormatter::fromTime(const SystemTime& time) const {
-  return fromTime(std::chrono::system_clock::to_time_t(time));
+  struct CachedTime {
+    size_t seconds_length;
+
+    struct Formatted {
+      std::string str;
+      SubsecondOffsets subsecond_offsets;
+      std::chrono::seconds epoch_time_seconds;
+    };
+    // A map is used to keep different formatted format strings at a given second.
+    std::unordered_map<std::string, const Formatted> formatted;
+  };
+  static thread_local CachedTime cached_time;
+
+  const std::chrono::nanoseconds epoch_time_ns =
+      std::chrono::duration_cast<std::chrono::nanoseconds>(time.time_since_epoch());
+
+  const std::chrono::seconds epoch_time_seconds =
+      std::chrono::duration_cast<std::chrono::seconds>(epoch_time_ns);
+
+  // Remove all the expired cached items.
```

Comment:
You could push this loop underneath the `if (item == cached_time.formatted.end())` scope, to avoid doing this on the hot path. We only need to GC every now and then, once per second will be fine.